nexus/mgs-updates/src/rot_updater.rs

```diff
@@ -203,23 +210,218 @@ impl SpComponentUpdateHelper for ReconfiguratorRotUpdater {
+                            update.sp_type,
+                            update.slot_id,
+                            &SpComponent::ROT.to_string(),
+                            expected_active_slot.toggled().to_u16(),
+                        )
+                        .await
+                })
+                .await;
+            let found_version = match found_inactive_caboose_result {
+                Ok(version) => {
+                    FoundVersion::Version(version.into_inner().version)
+                }
+                Err(error) => {
+                    if error_means_caboose_is_invalid(&error) {
+                        FoundVersion::MissingVersion
+                    } else {
+                        return Err(PrecheckError::from(error));
+                    }
+                }
+            };
+            match (&expected_inactive_slot_version, &found_version) {
+                // expected garbage, found garbage
+                (
+                    ExpectedVersion::NoValidVersion,
+                    FoundVersion::MissingVersion,
+                ) => (),
+                // expected a specific version and found it
+                (
+                    ExpectedVersion::Version(artifact_version),
+                    FoundVersion::Version(found_version),
+                ) if artifact_version.to_string() == *found_version => (),
+                // anything else is a mismatch
+                (ExpectedVersion::NoValidVersion, FoundVersion::Version(_))
+                | (ExpectedVersion::Version(_), FoundVersion::MissingVersion)
+                | (ExpectedVersion::Version(_), FoundVersion::Version(_)) => {
+                    return Err(PrecheckError::WrongInactiveVersion {
+                        expected: expected_inactive_slot_version.clone(),
+                        found: found_version,
+                    });
+                }
+            };
+
+            let (
+                active, pending_persistent_boot_preference, transient_boot_preference
+            ) = match &state.rot {
+                RotState::V2 {
+                    active,
+                    pending_persistent_boot_preference,
+                    transient_boot_preference,
+                    ..
+                }
+                | RotState::V3 {
+                    active,
+                    pending_persistent_boot_preference,
+                    transient_boot_preference,
+                    ..
+                } => (
+                    active, pending_persistent_boot_preference, transient_boot_preference
+                ),
+                RotState::CommunicationFailed { message } => {
+                    return Err(PrecheckError::RotCommunicationFailed { message: message.to_string() })
+                },
+            };
+
+            // If the active slot does not match the expected active slot, there is
+            // likely another update happening. Bail out.
+            if expected_active_slot != active {
```

Comment:
Could we move this check above where we fetch the caboose? It would save us the extra requests if it fails, but also it seems like "this isn't running the slot we thought" is one of the more fundamental checks to validate up front.